File: ToV/js/sim/feature-extractor.js
'use strict';
(function (root) {
  var engine = root.VillageEngine;
  if (!engine && typeof require === 'function') {
    engine = require('../engine.js');
    root.VillageEngine = engine;
  }
  if (!engine) throw new Error('feature-extractor: VillageEngine not found');

  var PHASES = ['SETUP', 'SEATS', 'NIGHT', 'MORNING', 'DAY', 'END'];
  var TEAMS = ['TOWN', 'MAFIA', 'NEUTRAL'];
  var MAX_DAYS = 25;

  function getRoleList() {
    return Object.keys(engine.ROLES);
  }

  function playerCountOf(state) {
    return (state && state.playerCount) || (state && state.players ? state.players.length : 0);
  }

  function getFeatureDimension(state) {
    var n = playerCountOf(state);
    var r = getRoleList().length;
    return 16 + r + n * (3 * r + 5);
  }

  function getCandidateDimension(state) {
    return 3 * getRoleList().length + 8;
  }

  function getVoteDimension(state) {
    return getFeatureDimension(state) + getCandidateDimension(state);
  }

  function getNightDimension(state) {
    return getFeatureDimension(state) + getCandidateDimension(state);
  }

  function getClaimDimension(state) {
    return getFeatureDimension(state);
  }

  function oneHot(roleList, roleId, out) {
    var idx = roleList.indexOf(roleId);
    for (var i = 0; i < roleList.length; i += 1) out.push(idx === i ? 1 : 0);
  }

  function pushRoleBlock(roleList, roleId, out) {
    oneHot(roleList, roleId, out);
    out.push(roleId == null ? 1 : 0);
  }

  function pushSheriffSlot(memories, pid, target, out) {
    var self = memories[pid] || {};
    var res;
    if (self.ownSheriffResults && self.ownSheriffResults[target] !== undefined) {
      res = self.ownSheriffResults[target];
    } else if (self.sheriffPublic && self.sheriffPublic[target] !== undefined) {
      res = self.sheriffPublic[target];
    }
    if (res === 'INNOCENT') out.push(0, 1, 0);
    else if (res === 'SUSPICIOUS') out.push(0, 0, 1);
    else out.push(1, 0, 0);
  }

  function pushConsigliereSlot(roleList, memories, pid, target, out) {
    var self = memories[pid] || {};
    var res = self.consigliereResults ? self.consigliereResults[target] : undefined;
    var idx = res === undefined || res === null ? -1 : roleList.indexOf(res);
    for (var i = 0; i < roleList.length; i += 1) out.push(idx === i ? 1 : 0);
  }

  function pushDeadSlot(roleList, state, memories, pid, target, out) {
    var self = memories[pid] || {};
    var known = self.deadRoles ? self.deadRoles[target] : undefined;
    if (known !== undefined && known !== null) {
      pushRoleBlock(roleList, known, out);
      return;
    }
    var entry = null;
    var graveyard = state.graveyard || [];
    for (var i = graveyard.length - 1; i >= 0; i -= 1) {
      if (graveyard[i].playerId === target) {
        entry = graveyard[i];
        break;
      }
    }
    if (!entry) {
      for (var j = 0; j <= roleList.length; j += 1) out.push(0);
      return;
    }
    if (entry.wasCleaned) {
      pushRoleBlock(roleList, null, out);
      return;
    }
    pushRoleBlock(roleList, entry.trueRole, out);
  }

  function extract(state, memories, playerId) {
    memories = memories || {};
    var roleList = getRoleList();
    var players = state.players || [];
    var n = playerCountOf(state);
    var self = memories[playerId] || {};
    var out = [];

    var phaseIdx = PHASES.indexOf(state.phase);
    for (var i = 0; i < PHASES.length; i += 1) out.push(phaseIdx === i ? 1 : 0);
    out.push((state.dayNumber || 0) / MAX_DAYS);
    var nightNumber = state.night && state.night.number !== undefined ? state.night.number : 1;
    out.push(nightNumber / MAX_DAYS);

    var alive = 0;
    players.forEach(function (p) { if (p.isAlive) alive += 1; });
    out.push(alive / n);

    var counts = { TOWN: 0, MAFIA: 0, NEUTRAL: 0 };
    players.forEach(function (p) {
      var role = engine.ROLES[p.assignedRole];
      var team = role ? role.team : 'NEUTRAL';
      if (counts[team] === undefined) team = 'NEUTRAL';
      counts[team] += 1;
    });
    out.push(counts.TOWN / n);
    out.push(counts.MAFIA / n);
    out.push(counts.NEUTRAL / n);

    oneHot(roleList, self.roleId, out);

    var selfTeam = self.side || self.team || 'NEUTRAL';
    for (var t = 0; t < TEAMS.length; t += 1) out.push(selfTeam === TEAMS[t] ? 1 : 0);
    out.push(players[playerId - 1] && players[playerId - 1].isAlive ? 1 : 0);

    players.forEach(function (p) {
      pushSheriffSlot(memories, playerId, p.id, out);
      pushConsigliereSlot(roleList, memories, playerId, p.id, out);
      pushRoleBlock(roleList, self.claims ? self.claims[p.id] : undefined, out);
      pushDeadSlot(roleList, state, memories, playerId, p.id, out);
    });

    return out;
  }

  function candidate(state, memories, playerId, candidateId) {
    var roleList = getRoleList();
    var players = state.players || [];
    var cand = players[candidateId - 1];
    var self = memories[playerId] || {};
    var out = [];
    out.push(cand && cand.isAlive ? 1 : 0);
    out.push(candidateId === playerId ? 1 : 0);
    out.push(self.mafiaMembers && self.mafiaMembers[candidateId] ? 1 : 0);
    pushSheriffSlot(memories, playerId, candidateId, out);
    pushConsigliereSlot(roleList, memories, playerId, candidateId, out);
    pushRoleBlock(roleList, self.claims ? self.claims[candidateId] : undefined, out);
    pushDeadSlot(roleList, state, memories, playerId, candidateId, out);
    return out;
  }

  root.FeatureExtractor = {
    PHASES: PHASES,
    TEAMS: TEAMS,
    MAX_DAYS: MAX_DAYS,
    getRoleList: getRoleList,
    getFeatureDimension: getFeatureDimension,
    getCandidateDimension: getCandidateDimension,
    getVoteDimension: getVoteDimension,
    getNightDimension: getNightDimension,
    getClaimDimension: getClaimDimension,
    extract: extract,
    candidate: candidate
  };
  if (typeof module !== 'undefined' && module.exports) {
    module.exports = root.FeatureExtractor;
  }
})(typeof window !== 'undefined' ? window : globalThis);

File: ToV/python/feature_extractor.py
import numpy as np

from config import PHASES, MAX_DAYS


class FeatureExtractor:
    def __init__(self, game_config):
        self.cfg = game_config
        self.base_dim = game_config.base_dim
        self.candidate_dim = game_config.candidate_dim
        self.vote_dim = game_config.vote_dim
        self.night_dim = game_config.night_dim
        self.claim_dim = game_config.claim_dim

    def _one_hot(self, idx, size):
        v = np.zeros(size, dtype=np.float64)
        v[idx] = 1.0
        return v

    def _role_one_hot(self, role_id, size):
        idx = self.cfg.role_index.get(role_id)
        if idx is None:
            return np.zeros(size, dtype=np.float64)
        return self._one_hot(idx, size)

    def _claim_one_hot(self, role_id, size):
        if role_id is None:
            return self._one_hot(self.cfg.role_count, size)
        idx = self.cfg.role_index.get(role_id)
        if idx is None:
            return self._one_hot(self.cfg.role_count, size)
        return self._one_hot(idx, size)

    @staticmethod
    def _grave_entry(state, target):
        for e in reversed(state.get('graveyard', [])):
            if e['playerId'] == target:
                return e
        return None

    def _sheriff_slot(self, memories, pid, target):
        r = memories.get(pid, {}).get('sheriff_results', {}).get(target)
        if r is None:
            r = memories.get(pid, {}).get('ownSheriffResults', {}).get(target)
        if r == 'INNOCENT':
            return np.array([0.0, 1.0, 0.0])
        if r == 'SUSPICIOUS':
            return np.array([0.0, 0.0, 1.0])
        return np.array([1.0, 0.0, 0.0])

    def _consigliere_slot(self, memories, pid, target):
        r = memories.get(pid, {}).get('consigliere_results', {}).get(target)
        if r is None:
            return np.zeros(self.cfg.role_count, dtype=np.float64)
        return self._role_one_hot(r, self.cfg.role_count)

    def _claim_slot(self, memories, pid, target):
        claims = memories.get(pid, {}).get('claims', {})
        return self._claim_one_hot(claims.get(target), self.cfg.role_count + 1)

    def _dead_slot(self, state, memories, pid, target):
        size = self.cfg.role_count + 1
        known = memories.get(pid, {}).get('dead_roles_known', {}).get(target)
        if known is not None:
            return self._claim_one_hot(known, size)
        entry = self._grave_entry(state, target)
        if entry is None:
            return np.zeros(size, dtype=np.float64)
        if entry.get('wasCleaned'):
            return self._one_hot(self.cfg.role_count, size)
        return self._claim_one_hot(entry.get('trueRole'), size)

    def _team_vector(self, team):
        v = np.zeros(3, dtype=np.float64)
        if team == 'TOWN':
            v[0] = 1.0
        elif team == 'MAFIA':
            v[1] = 1.0
        else:
            v[2] = 1.0
        return v

    def extract(self, state, memories, pid):
        cfg = self.cfg
        p = cfg.player_count
        feats = []

        phase = state.get('phase', 'NIGHT')
        pv = np.zeros(6)
        if phase in PHASES:
            pv[PHASES.index(phase)] = 1.0
        else:
            pv[5] = 1.0
        feats.append(pv)

        feats.append(np.array([state.get('dayNumber', 0) / MAX_DAYS]))
        feats.append(np.array([state.get('night', {}).get('number', 1) / MAX_DAYS]))

        alive = sum(1 for pl in state['players'] if pl['isAlive'])
        feats.append(np.array([alive / p]))

        tc = cfg.team_counts
        feats.append(np.array([tc.get('town', 0) / p, tc.get('mafia', 0) / p, tc.get('neutral', 0) / p]))

        mem = memories.get(pid, {})
        feats.append(self._role_one_hot(mem.get('role_id'), cfg.role_count))
        feats.append(self._team_vector(mem.get('team', 'NEUTRAL')))
        feats.append(np.array([1.0 if state['players'][pid - 1]['isAlive'] else 0.0]))

        for target in range(1, p + 1):
            feats.append(self._sheriff_slot(memories, pid, target))
            feats.append(self._consigliere_slot(memories, pid, target))
            feats.append(self._claim_slot(memories, pid, target))
            feats.append(self._dead_slot(state, memories, pid, target))

        return np.concatenate(feats)

    def _candidate(self, state, memories, pid, candidate):
        cfg = self.cfg
        feats = []
        cand = state['players'][candidate - 1]
        mem = memories.get(pid, {})
        feats.append(np.array([1.0 if cand['isAlive'] else 0.0]))
        feats.append(np.array([1.0 if candidate == pid else 0.0]))
        teammates = mem.get('mafia_teammates', set())
        feats.append(np.array([1.0 if candidate in teammates else 0.0]))
        feats.append(self._sheriff_slot(memories, pid, candidate))
        feats.append(self._consigliere_slot(memories, pid, candidate))
        feats.append(self._claim_slot(memories, pid, candidate))
        feats.append(self._dead_slot(state, memories, pid, candidate))
        return np.concatenate(feats)

    def vote_feature(self, state, memories, voter_id, candidate_id):
        return np.concatenate([self.extract(state, memories, voter_id),
                               self._candidate(state, memories, voter_id, candidate_id)])

    def night_feature(self, state, memories, actor_id, target_id):
        return np.concatenate([self.extract(state, memories, actor_id),
                               self._candidate(state, memories, actor_id, target_id)])

    def claim_feature(self, state, memories, player_id):
        return self.extract(state, memories, player_id)
